### PR TITLE
[unit-tests-only] add unit tests for edit collaborator components

### PR DESCRIPTION
**Pull request by dpakach**
## Description

[unit-tests-only] add unit tests for edit collaborator components

## Related Issue

- Related https://github.com/owncloud/web/issues/5234

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:

- [ ] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] ...

### CODE REVIEW

**Comment by ownclouders** [comment]
Results for <strong>oC10XGAPortrait1</strong> https://drone.owncloud.com/owncloud/web/18207/42/1
The following scenarios passed on retry:
- webUIDeleteFilesFolders/deleteFilesFolders.feature:59

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingInternalGroupsToRoot</strong> https://drone.owncloud.com/owncloud/web/18207/23/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+    etag: '"89128c0e8122002db57bd19c9ec33004"',
+    shareTypes: [],
+    downloadURL: '',
+    ownerDisplayName: displayNames[owner],
+    ownerId: owner,
+    canDownload: () => true,
+    isReceivedShare: () => true,
+    canBeDeleted: () => true,
+    canRename: () => true,
+    canShare: () => true
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js:188]: imo `type` is the only required attribute here, can you review this and use only required

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/CollaboratorsEditOptions.spec.js [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+      // set expiration date 10 days into the future
+      expirationDate.setDate(expirationDate.getDate() + 10)
+      // set expected expire time to the end of the day
+      expirationDate.setHours(23, 59, 59, 999)
+      // FIXME: timezone offset is manually added in thecode
+      expirationDate.setMinutes(expirationDate.getMinutes() + expirationDate.getTimezoneOffset())
+
+      const wrapper = getMountedWrapper({
+        user: 'user0',
+        expirationDate,
```
[the remark this answers is not in the document]
Comment (dpakach) [reply packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/CollaboratorsEditOptions.spec.js:175]: No it expects javascript Date object here

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+      expect(saveTrigger).not.toHaveBeenCalled()
+      await saveButton.trigger('click')
+      expect(saveTrigger).toHaveBeenCalledTimes(1)
+    })
+
+    it('should not display saving in progress indicator by default', () => {
+      const wrapper = getMountedWrapper({ user: 'user0' })
+
+      const spinner = wrapper.find(selectors.savingInProgress)
+      expect(spinner.exists()).toBe(false)
```
[the remark this answers is not in the document]
Comment (dpakach) [reply packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js:152]: why?
checking `toBeFalsy()` will also pass when the result is 0. if it returns anything other than false, I think we can assume its broken.

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingInternalGroupsToRoot</strong> https://drone.owncloud.com/owncloud/web/18287/23/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (kiranparajuli589) [reply packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js:152]: .exists() function only returns boolean.

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/CollaboratorsEditOptions.spec.js [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+      })
+    })
+
+    it('should set the expiration date according to the expirationDate prop passed to the component', () => {
+      const expirationDate = new Date()
+      // set expiration date 10 days into the future
+      expirationDate.setDate(expirationDate.getDate() + 10)
+      // set expected expire time to the end of the day
+      expirationDate.setHours(23, 59, 59, 999)
+      // FIXME: timezone offset is manually added in thecode
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/CollaboratorsEditOptions.spec.js:170]:
```suggestion
      // FIXME: timezone offset is manually added in the code
```
fixme??

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/CollaboratorsEditOptions.spec.js [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+  user2: 'User Two'
+}
+
+const userObj = name => {
+  return {
+    id: name,
+    additionalInfo: null,
+    name,
+    displayName: displayNames[name],
+    displayname: displayNames[name] // FIXME: some values use different property name for display name
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/CollaboratorsEditOptions.spec.js:281]: these are not used in the code anywhere, seem redundant.

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/CollaboratorsEditOptions.spec.js [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+      expect(wrapper.vm.enteredExpirationDate).toBe(expectedDate)
+    })
+  })
+})
+
+const storeOptions = data => {
+  let { user, owner, shareCapabilities } = data
+  if (!owner) {
+    owner = user
+  }
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/CollaboratorsEditOptions.spec.js:196]: owner seems unused after this. please remove if not needed.

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/CollaboratorsEditOptions.spec.js [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+      isOcis: () => false,
+      user: () => userObj(user),
+      capabilities: () => {
+        return {
+          files_sharing: shareCapabilities
+        }
+      }
+    }
+  }
+  return storeOpts
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/CollaboratorsEditOptions.spec.js:239]:
```suggestion
  return {
    state: {
      user: userObj(user)
    },
    modules: {
      Files: {
        namespaced: true,
        getters: {
          highlightedFile: () => {
            return { type: 'file' }
          }
        }
      }
    },
    getters: {
      isOcis: () => false,
      user: () => userObj(user),
      capabilities: () => {
        return {
          files_sharing: shareCapabilities
        }
      }
    }
  }
```

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js:152]: these all tests can use a same wrapper at top with spies predefined. no need to define separate wrapper for every test

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+  }
+}
+
+const storeOptions = data => {
+  let { user, owner } = data
+  if (!owner) {
+    owner = user
+  }
+
+  const storeOpts = {
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js:198]: you can return the object directly. this variable name seems odd to me

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+
+const userObj = name => {
+  return {
+    id: name,
+    additionalInfo: null,
+    name,
+    displayName: displayNames[name],
+    displayname: displayNames[name] // FIXME: some values use different property name for display name
+  }
+}
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js:272]: used in multiple files. can it be DRY?

**Review by pascalwengerter — CHANGES_REQUESTED** [review]
@dpakach please rebase, address reviews and refactor, would be glad to merge this one soon 🧗🏽‍♂️

**Comment by dpakach** [comment]
@pascalwengerter I have addressed the reviews can you review.

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingInternalGroupsToRoot</strong> https://drone.owncloud.com/owncloud/web/18825/23/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/CollaboratorsEditOptions.spec.js [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+    it('should show the label for the role', () => {
+      const roleLabel = wrapper.find(selectors.roleLabel)
+
+      expect(roleLabel.exists()).toBeTruthy()
+      expect(roleLabel.attributes()).toMatchObject({
+        class: 'oc-label',
+        for: 'files-collaborators-role-button'
+      })
+
+      expect(roleLabel.text()).toBe('Role')
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/CollaboratorsEditOptions.spec.js:43]: use snapshot tests to assert text
```suggestion
      expect(roleLabel).toMatchSnapshot()
```

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+describe('Edit Collaborator', () => {
+  describe('Error message', () => {
+    it('should display errors when form has errors', async () => {
+      const wrapper = getShallowMountedWrapper({ user: 'user0', collaborators: [] })
+
+      await wrapper.setData({ errors: 'some error' })
+      const errorMsg = wrapper.find(selectors.error)
+      expect(errorMsg.exists()).toBeTruthy()
+
+      expect(errorMsg.text()).toBe('some error')
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js:42]: use snapshot

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+  })
+
+  describe('Add People Button', () => {
+    it('should render edit collaborator hint', () => {
+      const wrapper = getShallowMountedWrapper({ user: 'user0', collaborators: [] })
+
+      const editHint = wrapper.find(selectors.editHint)
+      expect(editHint.exists()).toBe(true)
+
+      expect(editHint.text()).toBe('Editing share with User One')
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js:60]: use snapshot

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+      })
+    })
+  })
+
+  describe('Action Buttons', () => {
+    it('should render the cancel button', () => {
+      const wrapper = getShallowMountedWrapper({ user: 'user0' })
+      const cancelButton = wrapper.find(selectors.cancelButton)
+      expect(cancelButton.exists()).toBe(true)
+      expect(cancelButton.text()).toBe('Cancel')
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js:107]: use snapshot

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+    })
+
+    it('should render the save button', () => {
+      const wrapper = getShallowMountedWrapper({ user: 'user0' })
+      const saveButton = wrapper.find(selectors.saveButton)
+      expect(saveButton.exists()).toBe(true)
+      expect(saveButton.attributes().disabled).toBe('true')
+      expect(saveButton.text()).toBe('Save')
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js:115]: use snapshot

File: packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+
+      const wrapper = getMountedWrapper({ user: 'user0' })
+      const cancelButton = wrapper.find(selectors.cancelButton)
+
+      expect(cancelTrigger).not.toHaveBeenCalled()
+      await cancelButton.trigger('click')
+      expect(cancelTrigger).toHaveBeenCalledTimes(1)
+    })
+
+    it('save button should work', async () => {
```
Comment (kiranparajuli589) [inline packages/web-app-files/tests/unit/components/SideBar/Shares/Collaborators/EditCollaborator.spec.js:129]: little vague...please be more descriptive here

**Review by kiranparajuli589 — COMMENTED** [review]
lgtm

**Comment by individual-it** [comment]
blocked waiting for review @kulmann @pascalwengerter

**Comment by pascalwengerter** [comment]
Closed in favor of #6076, sorry for being unresponsive so long. The sidebar sharing components are receiving a major overhaul in #6039, I'll see what I can re-use from your work here - thanks!

**Pull request closed without merging**